src/Reflection/ReflectionConstant.php

```diff
@@ -0,0 +1,308 @@
+use Roave\BetterReflection\Reflection\StringCast\ReflectionConstantStringCast;
+use Roave\BetterReflection\Reflector\Exception\IdentifierNotFound;
+use Roave\BetterReflection\Reflector\Reflector;
+use Roave\BetterReflection\SourceLocator\Located\LocatedSource;
+use Roave\BetterReflection\Util\CalculateReflectionColum;
+use Roave\BetterReflection\Util\ConstantNodeChecker;
+use Roave\BetterReflection\Util\GetFirstDocComment;
+use function array_slice;
+use function count;
+use function explode;
+use function implode;
+use function substr_count;
+
+class ReflectionConstant implements Reflection, CoreReflector
+{
+    /** @var Reflector */
+    private $reflector;
+
+    /** @var Node\Stmt\Const_|Node\Expr\FuncCall */
+    private $node;
+
+    /** @var LocatedSource */
+    private $locatedSource;
+
+    /** @var NamespaceNode|null */
+    private $declaringNamespace;
+
+    /** @var int|null */
+    private $positionInNode;
+
+    /** @var int|float|mixed[]|string|bool|null const value */
+    private $value;
+
+    /** @var bool */
+    private $valueWasCached = false;
+
+    private function __construct()
+    {
+    }
+
+    /**
+     * Create a ReflectionConstant by name, using default reflectors etc.
+     *
+     * @throws IdentifierNotFound
+     */
+    public static function createFromName(string $constantName) : self
+    {
+        return (new BetterReflection())->constantReflector()->reflect($constantName);
+    }
+
+    /**
+     * Create a reflection of a constant
+     *
+     * @internal
+     *
+     * @param Node\Stmt\Const_|Node\Expr\FuncCall $node Node has to be processed by the PhpParser\NodeVisitor\NameResolver
+     */
+    public static function createFromNode(
+        Reflector $reflector,
+        Node $node,
+        LocatedSource $locatedSource,
+        ?NamespaceNode $namespace = null,
+        ?int $positionInNode = null
+    ) : self {
+        return $node instanceof Node\Stmt\Const_
+            ? self::createFromConstKeyword($reflector, $node, $locatedSource, $namespace, $positionInNode)
+            : self::createFromDefineFunctionCall($reflector, $node, $locatedSource);
+    }
+
+    private static function createFromConstKeyword(
+        Reflector $reflector,
+        Node\Stmt\Const_ $node,
+        LocatedSource $locatedSource,
+        ?NamespaceNode $namespace,
+        int $positionInNode
+    ) : self {
+        $constant                     = new self();
+        $constant->reflector          = $reflector;
+        $constant->node               = $node;
+        $constant->locatedSource      = $locatedSource;
+        $constant->declaringNamespace = $namespace;
+        $constant->positionInNode     = $positionInNode;
+
+        return $constant;
+    }
+
+    private static function createFromDefineFunctionCall(
+        Reflector $reflector,
+        Node\Expr\FuncCall $node,
+        LocatedSource $locatedSource
+    ) : self {
+        if (! ConstantNodeChecker::isValidDefineFunctionCall($node)) {
```

Comment:
Would be better to make this an assertion, such as `assertValidDefineFunctionCall()`, which can then throw with the given AST node. Currently, the exception has no context, which will make things hard to debug

Reply:
Done.